src/core/gui/menus/popoverMenus/PageTypeSelectionPopover.cpp

```diff
@@ -105,90 +127,308 @@ GtkWidget* createPreviewGrid(const std::vector<std::unique_ptr<PageTypeInfo>>& p
+    gtk_widget_set_margin_end(separator, 7);
+    return separator;
+}
+
 };  // namespace
 
-PageTypeSelectionPopover::PageTypeSelectionPopover(PageTypeHandler* typesHandler,
-                                                   PageBackgroundChangeController* controller, const Settings* settings,
-                                                   GtkApplicationWindow* win):
-        PageTypeSelectionMenuBase(typesHandler, settings, SELECTION_ACTION_NAME), controller(controller) {
+PageTypeSelectionPopover::PageTypeSelectionPopover(Control* control, Settings* settings, GtkApplicationWindow* win):
+        PageTypeSelectionMenuBase(control->getPageTypes(), settings, PAGETYPE_SELECTION_ACTION_NAME),
+        control(control),
+        controller(control->getPageBackgroundChangeController()),
+        settings(settings),
+        selectedPageSize(getInitiallySelectedPaperSize(settings)),
+        selectedOrientation(selectedPageSize ? selectedPageSize->orientation() : PaperOrientation::VERTICAL) {
     static_assert(is_action_namespace_match<decltype(win)>(G_ACTION_NAMESPACE));
 
+    PaperFormatUtils::loadDefaultPaperSizes(paperSizeMenuOptions);
+    // Add Special options
+    customPaperSizeIndex = paperSizeMenuOptions.size();
+    paperSizeMenuOptions.emplace_back(_("Custom"));
+    copyCurrentPaperSizeIndex = paperSizeMenuOptions.size();
+    paperSizeMenuOptions.emplace_back(_("Copy current page"));
+
+    this->controller->setPaperSizeForNewPages(this->selectedPageSize);
+
+    orientationAction.reset(createOrientationGAction(selectedOrientation), xoj::util::adopt);
+    g_action_map_add_action(G_ACTION_MAP(win), G_ACTION(orientationAction.get()));
+    g_simple_action_set_enabled(orientationAction.get(), selectedPageSize.has_value());
+    g_signal_connect(G_OBJECT(orientationAction.get()), "change-state", G_CALLBACK(changedOrientationSelectionCallback),
+                     this);
+
     g_action_map_add_action(G_ACTION_MAP(win), G_ACTION(typeSelectionAction.get()));
 
-    controller->setPageTypeForNewPages(this->selectedPT);
+    comboBoxChangeSelectionAction.reset(createPageSizeComboBoxChangeSelectionGAction(), xoj::util::adopt);
+    g_action_map_add_action(G_ACTION_MAP(win), G_ACTION(comboBoxChangeSelectionAction.get()));
+
+    pageSizeChangedAction.reset(createPageSizeChangedGAction(), xoj::util::adopt);
+    g_action_map_add_action(G_ACTION_MAP(win), G_ACTION(pageSizeChangedAction.get()));
+}
+template <bool changeComboBoxSelection>
+void PageTypeSelectionPopover::setSelectedPaperSize(const std::optional<PaperSize>& newPageSize) {
+    if (newPageSize != selectedPageSize) {
+        if constexpr (changeComboBoxSelection) {
+            ignoreComboBoxSelectionChange = true;
+            g_action_activate(G_ACTION(comboBoxChangeSelectionAction.get()),
+                              g_variant_new_uint32(getComboBoxIndexForPaperSize(newPageSize)));
+            ignoreComboBoxSelectionChange = false;
```

Comment:
This won't work because the GAction is actually just a helper to change the GtkComboBox selection. I need this helper because I can't save the GtkComboBox pointer from the createPopover method where these widgets are created as this method has to be const. But more about this problem in the comment thread for PageTypeSelectionPopover.cpp lines 200 to 201